lib/srv/discovery/kube_integration_watcher.go

```diff
@@ -0,0 +1,230 @@
+					continue
+				}
+
+				existingClusters, err := clt.GetKubernetesClusters(s.ctx)
+				if err != nil {
+					s.Log.WithError(err).Warn("Failed to get Kubernetes clusters from cache.")
+					continue
+				}
+
+				var newClusters []types.DiscoveredEKSCluster
+				mu.Lock()
+				for _, r := range resources {
+					newCluster, ok := r.(types.DiscoveredEKSCluster)
+					if !ok ||
+						enrollingClusters[newCluster.GetAWSConfig().Name] ||
+						slices.ContainsFunc(existingServers, func(c types.KubeServer) bool { return c.GetName() == newCluster.GetName() }) ||
+						slices.ContainsFunc(existingClusters, func(c types.KubeCluster) bool { return c.GetName() == newCluster.GetName() }) {
+
+						continue
+					}
+
+					newClusters = append(newClusters, newCluster)
+				}
+				mu.Unlock()
+
+				if len(newClusters) == 0 {
+					continue
+				}
+
+				agentVersion, err := s.getKubeAgentVersion(releaseChannels)
+				if err != nil {
+					s.Log.WithError(err).Warn("Could not get agent version to enroll EKS clusters")
+					continue
+				}
+
+				// When enrolling EKS clusters, client for enrollment depends on the region and integration used.
+				clustersByRegionAndIntegration := map[string]map[string][]types.DiscoveredEKSCluster{}
+				for _, c := range newClusters {
+					if _, ok := clustersByRegionAndIntegration[c.GetAWSConfig().Region]; !ok {
+						clustersByRegionAndIntegration[c.GetAWSConfig().Region] = map[string][]types.DiscoveredEKSCluster{}
+					}
+					clustersByRegionAndIntegration[c.GetAWSConfig().Region][c.GetIntegration()] =
+						append(clustersByRegionAndIntegration[c.GetAWSConfig().Region][c.GetIntegration()], c)
+				}
+
+				for region := range clustersByRegionAndIntegration {
+					for integration := range clustersByRegionAndIntegration[region] {
+						go s.enrollEKSClusters(region, integration, clustersByRegionAndIntegration[region][integration], agentVersion, &mu, enrollingClusters)
+					}
+				}
+
+			case <-s.ctx.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+func (s *Server) enrollEKSClusters(region, integration string, clusters []types.DiscoveredEKSCluster, agentVersion string, mu *sync.Mutex, enrollingClusters map[string]bool) {
+	mu.Lock()
+	for _, c := range clusters {
+		if _, ok := enrollingClusters[c.GetAWSConfig().Name]; !ok {
+			enrollingClusters[c.GetAWSConfig().Name] = true
+		}
+	}
+	mu.Unlock()
+	defer func() {
+		// Clear enrolling clusters in the end.
+		mu.Lock()
+		for _, c := range clusters {
+			delete(enrollingClusters, c.GetAWSConfig().Name)
+		}
+		mu.Unlock()
+	}()
+
+	// We sort input clusters into two batches - one that has Kubernetes App Discovery
+	// enabled, and another one that doesn't have it.
+	var batchedClusters = map[bool][]types.DiscoveredEKSCluster{}
+	for _, c := range clusters {
+		batchedClusters[c.GetKubeAppDiscovery()] = append(batchedClusters[c.GetKubeAppDiscovery()], c)
+	}
+	ctx, cancel := context.WithTimeout(s.ctx, time.Duration(len(clusters))*30*time.Second)
```

Comment:
Yep, it can overshoot estimated timeout because of parallelization, but since we can't predict it precisely anyway, I don't think it's worth it to create something more complicated here.